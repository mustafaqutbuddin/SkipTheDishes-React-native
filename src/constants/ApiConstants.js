
export const BASE_URL = "https://uk.api.just-eat.io/"  

//API TIMEOUT
export const TIMEOUT = 30000  //30 secs


//apis
export const GET_RESTAURANT_BY_POST_CODE = "restaurants/bypostcode/"