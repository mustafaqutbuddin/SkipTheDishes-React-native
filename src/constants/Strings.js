
//PLACEHOLDERS
export const POSTAL_CODE_PLACEHOLDER = "Enter postal code"
export const NO_DATA_PLACEHOLDER = "No records found."

//ERROR MESSAGES
export const NO_INTERNET_CONNECT = "No internet connection."
export const REQUEST_TIMED_OUT = "Something went wrong, Please try again later."
export const ERROR_TITLE = "Error!"



//HEADER TITLE
export const HEADER_HOME = "Home"



