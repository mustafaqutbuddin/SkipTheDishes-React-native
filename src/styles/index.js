import * as COLOR from './colors'
import * as RATIO from './mixins'

export {
    COLOR,
    RATIO
}