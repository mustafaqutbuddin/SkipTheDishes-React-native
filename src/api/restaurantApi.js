import * as API_CONSTANTS from '../constants/ApiConstants'
import apiClient from './config'



export const getRestaurantsByPostCode = async (postCode = 'ec4m', callback) => {
    try {
        postCode === '' ? 'ec4m' : postCode
        const response = await apiClient.get(API_CONSTANTS.GET_RESTAURANT_BY_POST_CODE + postCode)
        callback(response.data, undefined)
    } catch (error) {
        callback(undefined, error.message)
    }
}