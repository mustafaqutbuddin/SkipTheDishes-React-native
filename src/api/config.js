import axios from 'axios'
import * as API_CONSTANTS from '../constants/ApiConstants'
import { AlertHelper } from '../utils/AlertHelper'
import * as STRINGS from '../constants/Strings'

//axios config 
const apiClient = axios.create({
    baseURL: API_CONSTANTS.BASE_URL,
    timeout: API_CONSTANTS.TIMEOUT
})


apiClient.interceptors.response.use(
    config => config,
    (error) => {

        if (!error.response) {
            // no internet connectivity
            AlertHelper.show(AlertHelper.ALERT_TYPE.ERROR, STRINGS.ERROR_TITLE, STRINGS.NO_INTERNET_CONNECT)
        }
        // handle request timeout
        else if (error.code === 'ECONNABORTED' || error.response.status === 404) {
            AlertHelper.show(AlertHelper.ALERT_TYPE.ERROR, STRINGS.ERROR_TITLE, STRINGS.REQUEST_TIMED_OUT)
            console.log(`A timeout happend on url ${error.config.url}`)
            //handle bad request
        } else if (error.response.status === 400) {
            if (error.response.data.details) {
                AlertHelper.show(AlertHelper.ALERT_TYPE.ERROR, STRINGS.ERROR_TITLE, error.response.data.details[0].messages[0])
            } else {
                AlertHelper.show(AlertHelper.ALERT_TYPE.ERROR, STRINGS.ERROR_TITLE, error.response.data.messages[0])
            }
            console.log(`Error on this api === ${error.config.url}`)
        } else {
            //handle other erros
            AlertHelper.show(AlertHelper.ALERT_TYPE.ERROR, STRINGS.ERROR_TITLE, STRINGS.REQUEST_TIMED_OUT)
            console.log(`Error on this api === ${error.config.url}`)
        }
        return Promise.reject(error);
    },
)


export default apiClient