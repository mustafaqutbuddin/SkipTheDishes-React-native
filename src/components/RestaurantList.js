import React from 'react'
import { Text, StyleSheet, FlatList, View, Image } from 'react-native'
import { RATIO, COLOR } from '../styles'
import { Rating } from 'react-native-elements'
import { NO_DATA_PLACEHOLDER } from '../constants/Strings'


const RestaurantList = ({ data }) => {

    const emptyList = () => {
        return (
            <View style={styles.emptyList}>
                <Text>{NO_DATA_PLACEHOLDER}</Text>
            </View>
        )
    }

    const cuisineHeader = () => {
        return (
            <Text style={styles.cuisineHeader}>CuisineTypes: </Text>
        )
    }
    return (
        <FlatList
            showsVerticalScrollIndicator={false}
            data={data}
            ListEmptyComponent={emptyList}
            keyExtractor={(data, index) => index.toString()}
            contentContainerStyle={{ flexGrow: 1 }}
            renderItem={({ item }) => {
                return (
                    <View style={styles.itemContainer}>
                        <View style={{ flex: 0.65 }}>
                            <Image style={styles.image}
                                source={{ uri: item.LogoUrl }}
                                resizeMode='stretch'
                            />
                        </View>

                        <View style={{ flex: 0.35, justifyContent: 'center', marginHorizontal: RATIO.HEIGHT_PERCENTAGE(1) }}>
                            <Text style={styles.headerStyle}>{item.Name}</Text>

                            <View>
                                <FlatList style={{ marginVertical: RATIO.WIDTH_PERCENTAGE(2) }}
                                    scrollEnabled={false}
                                    horizontal
                                    ListHeaderComponent={cuisineHeader}
                                    data={item.CuisineTypes}
                                    keyExtractor={(data, index) => index.toString()}
                                    renderItem={({ item }) => {
                                        return (
                                            <Text style={styles.cuisines}>{item.Name}</Text>
                                        )
                                    }}
                                />
                            </View>
                            <View style={{ alignSelf: 'flex-end' }}>
                                <Rating imageSize={20}
                                    readonly
                                    startingValue={item.Rating.StarRating}
                                />
                            </View>
                        </View>
                    </View>
                )
            }}
        />
    )
}

const styles = StyleSheet.create({
    itemContainer: {
        marginHorizontal: RATIO.WIDTH_PERCENTAGE(2),
        height: RATIO.HEIGHT_PERCENTAGE(36),
        marginBottom: RATIO.HEIGHT_PERCENTAGE(2),
        backgroundColor: COLOR.WHITE,
        borderRadius: 10
    },

    image: {
        width: '100%',
        height: '100%',
        borderTopLeftRadius: 10,
        borderTopRightRadius: 10,
    },

    headerStyle: {
        fontSize: RATIO.scaleFont(1.7),
        fontWeight: 'bold'
    },

    addressStyle: {
        fontSize: RATIO.scaleFont(1.5),
        fontWeight: 'normal'
    },

    emptyList: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },

    cuisineHeader: {
        fontSize: RATIO.scaleFont(1.5),
        fontWeight: 'bold'

    },

    cuisines: {
        paddingLeft: RATIO.WIDTH_PERCENTAGE(2),
        fontSize: RATIO.scaleFont(1.5),
    }
})

export default RestaurantList
