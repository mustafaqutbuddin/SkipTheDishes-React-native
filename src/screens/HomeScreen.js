import React, { useState } from 'react'
import { StyleSheet, SafeAreaView, ActivityIndicator } from 'react-native'
import Input from '../components/Input'
import RestaurantList from '../components/RestaurantList'
import { RATIO } from '../styles'
import * as STRINGS from '../constants/Strings'
import useRestaurant from '../hooks/useRestaurant'



const HomeScreen = () => {
    //state variables
    const [postCode, setPostCode] = useState(undefined)
    const [loading, restaurants, handleOnSearch] = useRestaurant(postCode)

    return (

        <SafeAreaView style={styles.appContainer}>

            <Input
                containerStyle={styles.inputStyle}
                placeHolder={STRINGS.POSTAL_CODE_PLACEHOLDER}
                onChangeText={(text) => setPostCode(text)}
                onSubmit={handleOnSearch}
                value={postCode}
            />

            {loading
                ?
                <ActivityIndicator style={{ flex: 1, justifyContent: 'center', alignSelf: 'center' }}
                    size='large'
                    animating={loading}
                />
                :
                <RestaurantList data={restaurants} />
            }

        </SafeAreaView>
    )
}

const styles = StyleSheet.create({
    appContainer: {
        flex: 1,
        marginHorizontal: RATIO.WIDTH_PERCENTAGE(4)
    },

    inputStyle: {
        margin: RATIO.HEIGHT_PERCENTAGE(2),
    }
})

export default HomeScreen