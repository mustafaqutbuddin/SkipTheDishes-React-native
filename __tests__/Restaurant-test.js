import React from 'react'
import * as Api from '../src/api/restaurantApi'
import * as API_CONSTANTS from '../src/constants/ApiConstants'
import { HomeScreen } from '../src/screens'
import renderer, { act } from 'react-test-renderer'
import mockAxios from 'axios'
import RestaurantList from '../src/components/RestaurantList'

jest.useFakeTimers()

it('renders correctly', async () => {
    const tree = renderer
        .create(<HomeScreen />)
        .toJSON()
    expect(tree).toMatchSnapshot()
})



describe("testing api", () => {
    it('should call the api and get restaurant data', done => {
        const mockedResponse = {
            "Restaurants": [{
                "Id": 77574,
                "Name": "Swift Kebab House",
                "UniqueName": "swiftkebabhouse-e2"
            }]
        }
        //setup
        mockAxios.get.mockImplementationOnce(() => {
            return Promise.resolve({
                data: mockedResponse
            })
        })

        //work
        Api.getRestaurantsByPostCode('ec4m', (data, error) => {
            try {
                //assertions //expects
                expect(data).toEqual(mockedResponse)
                expect(mockAxios.get).toHaveBeenCalledTimes(1)
                expect(mockAxios.get).toHaveBeenCalledWith(API_CONSTANTS.GET_RESTAURANT_BY_POST_CODE + 'ec4m')
                done()
            } catch (error) {
                console.log(error)
                done(error)
            }
        })
    })
    
})

