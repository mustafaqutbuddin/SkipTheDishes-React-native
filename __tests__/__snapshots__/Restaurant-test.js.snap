// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "marginHorizontal": 30,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#F05E40",
        "borderRadius": 40,
        "height": 86.5,
        "justifyContent": "center",
        "margin": 26.5,
        "width": 637.5,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      editable={true}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Enter postal code"
      placeholderTextColor="rgba(255,255,255, 0.6)"
      rejectResponderTermination={true}
      style={
        Object {
          "color": "#fff",
          "fontSize": 29,
          "height": "100%",
          "padding": 0,
          "paddingHorizontal": 30,
          "width": "100%",
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "alignSelf": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  />
</RCTSafeAreaView>
`;
